TS.fsx

```diff
@@ -117,32 +117,55 @@ let matchSingleParamMethodSignature (m: Browser.Method) expectedMName expectedMT
 /// Emit overloads for the createElement method
 let EmitCreateElementOverloads (m: Browser.Method) =
     if matchSingleParamMethodSignature m "createElement" "Element" "string" then
-        for e in tagNameToEleName do
-            if iNameToIDependList.ContainsKey e.Value && Seq.contains "HTMLElement" iNameToIDependList.[e.Value] then
-                Pt.printl "createElement(tagName: \"%s\"): %s;" e.Key e.Value
+        Pt.printl "createElement<K extends keyof HTMLElementTagNameMap>(tagName: K): HTMLElementTagNameMap[K];"
         Pt.printl "createElement(tagName: string): HTMLElement;"
 
 /// Emit overloads for the getElementsByTagName method
 let EmitGetElementsByTagNameOverloads (m: Browser.Method) =
     if matchSingleParamMethodSignature m "getElementsByTagName" "NodeList" "string" then
-        for e in tagNameToEleName do
-            Pt.printl "getElementsByTagName(%s: \"%s\"): NodeListOf<%s>;" m.Params.[0].Name (e.Key.ToLower()) e.Value
+        Pt.printl "getElementsByTagName<K extends keyof ElementListTagNameMap>(%s: K): ElementListTagNameMap[K];" m.Params.[0].Name
         Pt.printl "getElementsByTagName(%s: string): NodeListOf<Element>;" m.Params.[0].Name
 
 /// Emit overloads for the querySelector method
 let EmitQuerySelectorOverloads (m: Browser.Method) =
     if matchSingleParamMethodSignature m "querySelector" "Element" "string" then
-        for e in tagNameToEleName do
-            Pt.printl "querySelector(selectors: \"%s\"): %s | null;" (e.Key.ToLower()) e.Value
+        Pt.printl "querySelector<K extends keyof ElementTagNameMap>(selectors: K): ElementTagNameMap[K] | null;"
         Pt.printl "querySelector(selectors: string): Element | null;"
 
 /// Emit overloads for the querySelectorAll method
 let EmitQuerySelectorAllOverloads (m: Browser.Method) =
     if matchSingleParamMethodSignature m "querySelectorAll" "NodeList" "string" then
-        for e in tagNameToEleName do
-            Pt.printl "querySelectorAll(selectors: \"%s\"): NodeListOf<%s>;" (e.Key.ToLower()) e.Value
+        Pt.printl "querySelectorAll<K extends keyof ElementListTagNameMap>(selectors: K): ElementListTagNameMap[K];"
         Pt.printl "querySelectorAll(selectors: string): NodeListOf<Element>;"
 
+let EmitHTMLElementTagNameMap () =
+    Pt.printl "interface HTMLElementTagNameMap {"
+    Pt.increaseIndent()
+    for e in tagNameToEleName do
+        if iNameToIDependList.ContainsKey e.Value && Seq.contains "HTMLElement" iNameToIDependList.[e.Value] then
+            Pt.printl "\"%s\": %s;" (e.Key.ToLower()) e.Value
+    Pt.decreaseIndent()
+    Pt.printl "}"
+    Pt.printl ""
+
+let EmitElementTagNameMap () =
+    Pt.printl "interface ElementTagNameMap {"
+    Pt.increaseIndent()
```

Comment:
Aha now I remembder, the `EmitHTMLElementTagNameMap` only prints tag elements that either extends or implements the `HTMLElement` type.